lib/puppet/util/run_mode.rb

```diff
@@ -14,7 +14,7 @@ def self.[](name)
         if Puppet::Util::Platform.windows?
           @run_modes[name] ||= WindowsRunMode.new(name)
         else
-          @run_modes[name] ||= UnixRunMode.new(name)
+          @run_modes[name] ||= LinuxFHSRunMode.new(name)
```

Comment:
I wonder how we can signal this runmode. Can install.rb do something about this? I don't like the idea of it patching code but maybe it should?

Reply:
I'm still thinking about this. Ideally we should maintain the same path structure as before when running puppet as a gem or in development mode with bundler.

Honestly patching might be a cleaner solution... it's a bit weird because paths are overridable in `install.rb` but they are still hardcoded here. We should find a way to keep them in sync.

Reply:
I agree. Another option is to write a file (like an ini file or whatever) with the paths as install.rb defined them somewhere in the gem directory and load it it at runtime.

Reply:
Due to the way run mode and settings are intertwined, we can't specify the run mode in puppet.conf. You could potentially load paths from a different ini file, unrelated to settings.... Or just patch.

Reply:
Interesting, FreeBSD has its own paths (mostly the same but under `/usr/local/`).  Patching this line when packaging on FreeBSD is indeed an option if providing a FreeBSDRunMode class is acceptable, and that would make my life easier :wink:

Reply:
I am wondering if `__FILE__` could be used to guess what class to use.  If it match `/opt/puppetlabs/puppet/lib/.*` we have an AIO Puppet, and otherwise some OS-detection just like windows do might help?

Reply:
> Due to the way run mode and settings are intertwined, we can't specify the run mode in puppet.conf. You could potentially load paths from a different ini file, unrelated to settings.... Or just patch.

That was the path I investigated: somehow do that in `install.rb` (hence the clean up PR you saw). I still don't quite know what it should be. It's almost like `install.rb` should write out the answers and load them.

Or what I also considered: specify a runmode to `install.rb` and load the answers from `RunMode`.

I welcome any suggestions what would be preferred.

> Interesting, FreeBSD has its own paths (mostly the same but under `/usr/local/`). Patching this line when packaging on FreeBSD is indeed an option if providing a FreeBSDRunMode class is acceptable, and that would make my life easier 😉

I had hoped to use this to unify the various Linux distros, but if this makes it easier for BSDs then that's a nice bonus.

Reply:
On the `install.rb` path I came up with https://github.com/puppetlabs/puppet/pull/9490. By making it use `RunMode` we can then simplify patch the RunMode to get all the correct paths.